Jason Hendrawan - 2501963040

In [ ]:
!pip install -U scikit-learn

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv('/content/Dataset_Classification.csv')

In [ ]:
df.head()

x1        x2        x3        x4        x5        x6        x7  \
0 -1.032690  2.096221  0.384194 -1.214555  2.432575  1.069743 -1.076878   
1 -0.509585 -0.896533 -0.465612  0.033544 -0.781463 -0.934908  0.061976   
2 -1.032690 -0.896533  1.243594 -1.101241 -0.781463 -0.934908 -0.355857   
3 -1.024914  1.367049  0.014966 -1.141605  1.102526  0.197138 -0.921322   
4  1.585219 -0.896533 -0.782318  0.692780 -0.781463  0.732274 -0.113010   

         x8        x9       x10  ...      x248      x249      x250      x251  \
0  2.140959  0.403971 -1.039528  ...  2.423807  0.833363 -0.932956  2.047133   
1 -0.712155 -0.929289 -0.624238  ... -0.783508 -0.877666  0.084089 -0.718668   
2 -0.020836 -0.726172 -1.039528  ... -0.783508 -0.851601 -0.481456  0.041508   
3  1.064377  0.235659 -0.973558  ...  1.228601  0.089344 -0.932956  1.093751   
4 -1.048795  1.248580  1.636419  ... -0.783508  0.902548  0.000080 -1.049062   

       x252      x253      x254      x255      x256  label  
0  0.089361 -1.062281  2.080691  0.412019 -1.244573      3  
1 -0.965767 -0.567989 -0.911335 -0.415529  0.167377      3  
2 -0.623651 -1.062281 -0.911335  1.288949 -1.201469      0  
3  0.017282 -0.920086  1.336901  0.012367 -0.990998      3  
4  1.518648  1.607938 -0.911335 -0.824106  0.663292      0  

[5 rows x 257 columns]

In [ ]:


import matplotlib.pyplot as plt
import seaborn as sns

corr = df.corr()

plt.figure(figsize=(20, 20))
sns.heatmap(corr, annot= True, cmap = 'coolwarm' )
plt.show()

# a. EDA

In [ ]:
df.dtypes

x1       float64
x2       float64
x3       float64
x4       float64
x5       float64
          ...   
x253     float64
x254     float64
x255     float64
x256     float64
label      int64
Length: 257, dtype: object

hasil dari df.dtypes diatas adalah bentuk data dari data csv nya jadi untuk x1-x256 itu berbentuk float dan label berbentuk integer

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3264 entries, 0 to 3263
Columns: 257 entries, x1 to label
dtypes: float64(256), int64(1)
memory usage: 6.4 MB


untuk df.info ini menjelaskanbahwa ada 257 column dari x1 sampai ke label, ada 256 data float dan 1 integer, dan ada total 3264 row

In [ ]:
df.isna().sum()

x1       0
x2       0
x3       0
x4       0
x5       0
        ..
x253     0
x254     0
x255     0
x256     0
label    0
Length: 257, dtype: int64

df.isna().sum() itu berguna untuk mencari jumlah NaN data pada tiap column di data ini, dari x1-label memiliki hasil 0, yang berarti tidak ada data NaN pada data ini

# b. praproses

In [ ]:
X = df.drop(columns=['label'], axis = 1)
y = df['label']

df.dropna(axis = 1, inplace = True)

In [ ]:
from sklearn.preprocessing import StandardScaler


scaler = StandardScaler()
X = scaler.fit_transform(X)
X

array([[-1.00610712,  1.88059418,  0.39499165, ...,  1.89305181,
         0.44488926, -1.18293629],
       [-0.43670779, -0.95217097, -0.50864155, ..., -0.93985297,
        -0.4363631 ,  0.18437406],
       [-1.00610712, -0.95217097,  1.30882672, ..., -0.93985297,
         1.37872892, -1.14119447],
       ...,
       [-0.76155071,  0.84353978,  1.15908573, ...,  0.84038672,
         1.07603927, -0.83681362],
       [ 1.7685804 , -0.95217097, -1.38092082, ..., -0.93985297,
        -1.34149895,  0.52987896],
       [-0.1982883 ,  0.51148773, -0.22580351, ...,  0.50176234,
        -0.28914506,  0.65634939]])

pertama saya buat variabel X dan y yang berisi y itu label dan x sisanya, kemudian dalam praproses saya drop semua NaN, walaupun tidak ada data NaN yang terdeteksi sebelumnya tapi saya tetap masukan aja untuk memastikan tidak ada NaN yang dapat merusak.

kemudian saya melakukan standarisasi dengan standard scaler, hal ini saya lakukan agar distribusinya terpusat pada 0 dan standar deviasinya 1, namun standarisasi ini kurang baik jika bukan gaussian distribution.

# C. 2 metode machine learning klasifikasi

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [ ]:
from sklearn.metrics import classification_report

model 1: KNN

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
neigh = KNeighborsClassifier(n_neighbors=5)
neigh.fit(X_train, y_train)

y_pred_knn = neigh.predict(X_test)


model 2: Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier(n_estimators=10)

rfc.fit(X_train, y_train)
y_pred_rfc = rfc.predict(X_test)


saya memilih kedua model ini karena nearest neighbor dan random forest bekerja baik dan memberikan accuracy terbesar diantara model" lain yang saya coba seperti adaboost, svm, logistic regression, dll.

KNN ada classifier yang konsisten dengan distribusiapapun, jadi jika diberikan banyak training, knn tetap memberikan hasil optimal

random tree juga saya pilih baik dalam menangani overfitting

#d. Metric evaluation

In [ ]:
from sklearn.metrics import accuracy_score, f1_score, recall_score, precision_score, classification_report

In [ ]:
print(accuracy_score(y_test, y_pred_knn))
print(f1_score(y_test, y_pred_knn, average = 'weighted'))
print(recall_score(y_test, y_pred_knn, average = 'weighted'))
print(precision_score(y_test, y_pred_knn, average = 'weighted'))

print(classification_report(y_test, y_pred_knn))

0.7010204081632653
0.6981352665835703
0.7010204081632653
0.6971650891858007
              precision    recall  f1-score   support

           0       0.65      0.65      0.65       270
           1       0.68      0.62      0.65       295
           2       0.62      0.60      0.61       139
           3       0.81      0.89      0.84       276

    accuracy                           0.70       980
   macro avg       0.69      0.69      0.69       980
weighted avg       0.70      0.70      0.70       980



In [ ]:

print(accuracy_score(y_test, y_pred_rfc))
print(f1_score(y_test, y_pred_rfc, average = 'weighted'))
print(recall_score(y_test, y_pred_rfc, average = 'weighted'))
print(precision_score(y_test, y_pred_rfc, average = 'weighted'))

print(classification_report(y_test, y_pred_rfc))

0.7459183673469387
0.7450818385891587
0.7459183673469387
0.7445937289651047
              precision    recall  f1-score   support

           0       0.71      0.70      0.71       270
           1       0.70      0.69      0.70       295
           2       0.75      0.73      0.74       139
           3       0.82      0.86      0.84       276

    accuracy                           0.75       980
   macro avg       0.75      0.74      0.74       980
weighted avg       0.74      0.75      0.75       980



dari kedua model saya dapatkan hasil yang cukup mirip dengan perbedaan sebesar 0.03 - 0.05 pada akurasi, f1 score, recall, precision. tetapi dalam setiap percobaan Random forest selalu menghasilkan akurasi tertinggi sehingga menurut saya random forest adalah model terbaik untuk kasus ini.